storage/src/mmr/mutable.rs

```diff
@@ -0,0 +1,512 @@
+                if i == 0 {
+                    // Record 0 is a placeholder and not an actual deletion.
+                    continue;
+                }
+                let Some(old_location) = store.snapshot.remove(&record.key) else {
+                    error!("record deletes non-existent key");
+                    return Err(Error::InvalidUpdate);
+                };
+                &store.records[old_location]
+            };
+
+            if old_record.key != record.key {
+                error!("record.key doesn't match old record");
+                return Err(Error::InvalidUpdate);
+            }
+
+            let bit_offset = leaf_pos_to_num(old_record.tree_pos);
+            assert!(store.is_active.get_bit(bit_offset));
+            store.is_active.set_bit(hasher, bit_offset, false);
+        }
+
+        assert_eq!(store.is_active.bit_count() as usize, store.records.len());
+
+        Ok(store)
+    }
+
+    /// Return a digest of the `key` `value` pair.
+    pub fn key_value_hash(hasher: &mut H, key: &K, value: &V) -> H::Digest {
+        hasher.update(key);
+        hasher.update(value);
+        hasher.finalize()
+    }
+
+    /// Get the value of `key` in the store, or None if it has no value.
+    pub fn get(&self, key: &K) -> Option<&V> {
+        let pos = self.snapshot.get(key)?;
+        let record = &self.records[*pos];
+
+        Some(&record.value)
+    }
+
+    /// Set the value of `key` to `value`.
+    pub fn update(&mut self, hasher: &mut H, key: K, value: V) {
+        // Update the updates tree
+        let record_digest = Self::key_value_hash(hasher, &key, &value);
+        let tree_pos = self.updates.add(hasher, &record_digest);
+
+        // Set the is_active bit of the new record to 1.
+        self.is_active.append(hasher, true);
+
+        // Update the snapshot.
+        let new_location = self.records.len();
+        if let Some(old_location) = self.snapshot.insert(key.clone(), new_location) {
+            // Flip the is_active bit of the old record to 0.
+            let record = &self.records[old_location];
+            assert_eq!(record.key, key);
+            let bit_offset = leaf_pos_to_num(record.tree_pos);
+            assert!(self.is_active.get_bit(bit_offset));
+            self.is_active.set_bit(hasher, bit_offset, false);
+        }
+
+        // Append the new record to the records vector.
+        let new_record = UpdateRecord {
+            tree_pos,
+            key,
+            value,
+        };
+        self.records.push(new_record);
+        assert_eq!(self.records.len(), self.is_active.bit_count() as usize);
+    }
+
+    /// Delete `key` and its value from the store. Panic if `key` has no value.
+    pub fn delete(&mut self, hasher: &mut H, key: K) {
+        // Remove the key from the snapshot.
+        let old_location = match self.snapshot.remove(&key) {
+            Some(location) => location,
+            None => panic!("key has no value {}", key),
```

Comment:
I think it is ok to support deleting keys that don't exist (just return here)?

Reply:
I suspect it will simplify application logic

Reply:
ok changed & added tests to confirm deleting keys that don't exist is a no-op. I usually prefer to force users to be precise in their usage of an API but acknowledge this could be overkill :)

Reply:
Definitely appreciate the intention, however, have already found myself doing things like this in the code (to avoid "tracking what I've done" in a sense).

On the user-level, if we had some raw KV application and allowed users to issue delete txs, this behavior would require us to do a read before delete (to ensure the key existed first).